### ISSUE TITLE
`cnb_runtime` requires Error to be the same for both `build` and `detect`

### ISSUE THREAD

**Issue by hone** [Malax/libcnb.rs#78]
With the change to the single binary and the `cnb_runtime` method, both `build` and `detect` must return the same typed Error. I'm not sure if this is intentional.

`detect` is usually a much simpler method where you may want to restrict the Error enum returned here.

**Comment by Malax** [comment]
Using the same type for errors in detect and build was intentional to reduce the amount of type annotations everywhere. 

The idea is to have an error type that is specific to the buildpack. If a user of this library wants to make a distinction, they can still do something like this:

```rust
enum MyBuildpackError {
    DetectError(MyBuildpackDetectError),
    BuildError(MyBuildpackBuildError),
}

enum MyBuildpackDetectError {
    // ...
}

enum MyBuildpackBuildError {
    // ...
}
```

But I think this isn't necessary for the typical buildpack. If we think about a Maven buildpack, it would read `pom.xml` in `detect` but also reads it during `build`, allowing the concrete error for this to be shared across `detect` and `build`.

It would make things a little bit easier in testing, but not by much. When testing fail states for `detect`, we have to match against the expected error value anyway which doesn't get more verbose or complex if the error type has more potential values.

Personally, I think less type annotations are a good thing and the negative impact of this design aren't too bad. Thoughts?

**Comment by hone** [comment]
I don't feel super strongly about it, but ran into it when reviewing the Procfile buildpack that Richard wrote. Detect had 0 error states and Build had all of them.

I thought you loved type annotations reflecting the actual state of the world. :)

**Comment by Malax** [comment]
> I thought you loved type annotations reflecting the actual state of the world. :)

Yeah, but within reason. 😄  If we need to plaster more generic type annotations everywhere to meet an idealistic goal without much other use... 🤷🏻 

@hone Are you fine with me closing this as "wontfix"?

**Comment by edmorley** [comment]
I was wondering if this would be easier after #129, since presumably this would instead be implemented via separate `BuildError` and `DetectError` associated types on `Buildpack`. However I see there are codepaths that require an error type that are not build/detect specific, so that wouldn't work as is. eg:
https://github.com/Malax/libcnb.rs/blob/822a2cf4d622f7c4ba8a7ab44341644354faf525/libcnb/src/runtime.rs#L29

As such, it still seems this may not be worth the added complexity, and that we should close as wontfix?

**Comment by Malax** [comment]
> As such, it still seems this may not be worth the added complexity, and that we should close as wontfix?
👍🏻  from me.
